pyteal/ast/router.py

```diff
@@ -339,52 +339,83 @@ def wrap_handler(
-                last_arg_specs_grouped = arg_type_specs[METHOD_ARG_NUM_CUTOFF - 1 :]
-                arg_type_specs = arg_type_specs[: METHOD_ARG_NUM_CUTOFF - 1]
-                last_arg_spec = abi.TupleTypeSpec(*last_arg_specs_grouped)
-                arg_type_specs.append(last_arg_spec)
-
-            arg_abi_vars: list[abi.BaseType] = [
-                type_spec.new_instance() for type_spec in arg_type_specs
+
+            # All subroutine arg values, initialize here and use below instead of
+            # creating new instances on the fly so we dont have to think about splicing
+            # back in the transaction types
+            arg_vals = [typespec.new_instance() for typespec in arg_type_specs]
+
+            # Only args that appear in app args
+            app_arg_vals: list[abi.BaseType] = [
+                ats for ats in arg_vals if not isinstance(ats, abi.Transaction)
             ]
-            decode_instructions: list[Expr] = [
-                arg_abi_vars[i].decode(Txn.application_args[i + 1])
-                for i in range(len(arg_type_specs))
+
+            # assign to a var here since we modify app_arg_vals later
+            tuplify = len(app_arg_vals) > METHOD_ARG_NUM_CUTOFF
+
+            # only transaction args (these are omitted from app args)
+            txn_arg_vals: list[abi.BaseType] = [
+                ats for ats in arg_vals if isinstance(ats, abi.Transaction)
             ]
 
-            if handler.subroutine.argument_count() > METHOD_ARG_NUM_CUTOFF:
-                tuple_arg_type_specs: list[abi.TypeSpec] = cast(
-                    list[abi.TypeSpec],
-                    handler.subroutine.expected_arg_types[METHOD_ARG_NUM_CUTOFF - 1 :],
+            # Tuple-ify any app args after the limit
+            if tuplify:
+                last_arg_specs_grouped: list[abi.TypeSpec] = [
+                    t.type_spec() for t in app_arg_vals[METHOD_ARG_NUM_CUTOFF - 1 :]
+                ]
+                app_arg_vals = app_arg_vals[: METHOD_ARG_NUM_CUTOFF - 1]
+                app_arg_vals.append(
+                    abi.TupleTypeSpec(*last_arg_specs_grouped).new_instance()
                 )
-                tuple_abi_args: list[abi.BaseType] = [
-                    t_arg_ts.new_instance() for t_arg_ts in tuple_arg_type_specs
+
+            # decode app args
+            decode_instructions: list[Expr] = [
+                app_arg.decode(Txn.application_args[idx + 1])
+                for idx, app_arg in enumerate(app_arg_vals)
+            ]
+
+            # "decode" transaction types by setting the relative index
+            if len(txn_arg_vals) > 0:
+                txn_relative_pos = len(txn_arg_vals)
+                txn_decode_instructions: list[Expr] = [
+                    cast(abi.Transaction, arg_val).set(
+                        Txn.group_index() - Int(txn_relative_pos - idx)
+                    )
+                    for idx, arg_val in enumerate(txn_arg_vals)
+                ]
```

Comment:
this part is somewhat tricky, could you explain more on relative index and group index relation?

Reply:
The transactions should appear in the group in the order they're specified in the method signature and should be relative to the current transaction.

ex:
`doit(axfer,pay,appl)`
would be 4 transactions
current_idx-3 = axfer
current_idx-2 = pay
current_idx-1 = appl
current_idx-0  = the txn that triggered the current eval

since we're iterating in order of the txns appearance in the args we just subtract the current index from the total length to get the offset.

Do you want this explanation in the source file as well as a comment?

Reply:
great, thanks for explaining and that makes sense.

I think the explanation can live as a chunk of comment in router code, that helps a lot.